### PR TITLE
seqr models

### PR DESCRIPTION
**Pull request by knguyen142**
tldr;
Fun stuff, introducing matrix table schemas
- clearly define annotation schema with fields and their values
- centralized so the logic is clearer
- express dependencies between annotations
- allow different types of schema to share annotation definitions but branch
- more easy to do validations and ensure correctness (TODO, with hail and ES types)

**----- The sell ----**
Are you like me, folks? Do you find yourself having a hard time tracking all of the seqr annotations across a long pipeline and from many different sources? Sometimes different tables storing different data should have slightly different annotations while sharing common ones, but the branching logic is just scattered. Other times, annotations depend on other annotations, so you have to order them. Introducing schema models!
**-----------**

Example:

```
        class TestSchema(BaseMTSchema):

            @mt_annotation()
            def a(self):
                return 0
            
            # depend on a
            @mt_annotation(fn_require='a')
            def b(self):
                return self.mt.a + 1

            # by default, use function name as annotation name but can override with `annotation=` 
            @mt_annotation(annotation='c', fn_require='a')
            def c_1(self):
                return self.mt.a + 2
```

Usage:
```
schema = TestSchema(mt)
schema.b() # annotates() `a` then`b` because of `fn_require='a'`.
schema.c() # will not call `a` again
mt = schema.select_annotated_mt() # return a matrix table with rows of only selected annotations.
## $ PROFIT. Also option to annotate all at once.
```

For our use case, I found it hard to track down all of the annotations across `variant` and `sv` types, then derived fields going off of vep and sortedtranscipt field, and later the reference data as well as annotations from genotypes. Having a clearly defined schema seems like it would make validation and code readability/maintainability.

### CODE REVIEW

**Comment by tpoterba** [comment]
I really like this as infrastructure for dealing with complex dependent schemas. 

Maybe we could use for building the [annotation database](https://hail.is/docs/0.1/annotationdb.html) in 0.2 -- something we have been intending to do for quite a while!

**Comment by knguyen142** [comment]
Thanks Tim! Yeah let me know how I can help, I found it very useful in modeling the annotation relationships and thought others might as well. I was even thinking of an interface where schema definitions can model transitions, like `SchemaA().load('<pathA>').transition_to(SchemaB()).write('<pathB>')` so you have general ways of going from any (valid) schema to SchemaB. But it is a more software engineering way to approach the problem than computational biologist. 

Just checking, but `mt.annotate_rows(A=1).annotate_rows(B=2)` and `mt.annotate_rows(A=1, B=2)` are equivalent in performance since we're just modifying the DAG and the only difference is creating trivial python objects? When I get a chance to get familiar with the codebase, I'll be able to dig through the code to figure it out.

File: luigi_pipeline/lib/model/base_mt_schema.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    def mt_prop_wrapper(func):
+        annotation_name = annotation or func.__name__
+
+        @wraps(func)
+        def wrapper(self, *args, **kwargs):
+            # Called already.
+            if wrapper.mt_prop_meta['annotated'] > 0:
+                return self
+            if fn_require:
+                getattr(self, fn_require.__name__)()
```
Comment (nawatts) [inline luigi_pipeline/lib/model/base_mt_schema.py]: This seems unintuitive given that it looks like an arbitrary function can be passed as a requirement but it must actually be a method of the same schema class.

**Comment by tpoterba** [comment]
> Just checking, but mt.annotate_rows(A=1).annotate_rows(B=2) and mt.annotate_rows(A=1, B=2) are equivalent in performance since we're just modifying the DAG and the only difference is creating trivial python objects? When I get a chance to get familiar with the codebase, I'll be able to dig through the code to figure it out.

The latter would be way faster to execute right now (some infrastructure changes will improve this so that an `annotate` doesn't require copying all the existing fields, just a pointer to them). However, our optimizer should rewrite the the former, split annotations into literally the same intermediate representation as the latter.

File: luigi_pipeline/lib/model/base_mt_schema.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+            @mt_annotation(fn_require=a)
+            def b(self):
+                return self.a + 1
+
+            @mt_annotation(annotation='c', fn_require=a)
+            def c_1(self):
+                return self.a + 2
+
+    `TestSchema(mt).b().c_1().select_annotated_mt()` will annotate with {'a': 0, 'b': 1, 'c': 2}
```
Comment (nawatts) [inline luigi_pipeline/lib/model/base_mt_schema.py]: This API seems a bit unwieldy. To me, `Schema(mt).foo()` looks like it would return an expression for the `foo` annotation vs modifying the instance's internal MatrixTable.

It also seems unlikely that someone creating a schema would ever want some but not all of the annotation fields, so the common use case would be `TestSchema(mt).annotate_all().mt`? Would something like `TestSchema.annotate_mt(mt)` work?

File: luigi_pipeline/lib/model/base_mt_schema.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    def annotate_all(self):
+        """
+        Iterate over all annotation functions and call them on the instance.
+        :return: instance object
+        """
+        for atn_fn in self.all_annotation_fns():
+            getattr(self, atn_fn[0])()
+        return self
+
+    def select_annotated_mt(self):
```
Comment (nawatts) [inline luigi_pipeline/lib/model/base_mt_schema.py]: Should this call `annotate_all` to ensure annotations have been applied?

File: luigi_pipeline/lib/model/base_mt_schema.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            if fn_require:
+                getattr(self, fn_require)()
+
+            # Annotate and retiurn instance for chaining.
+            self.mt = self.mt.annotate_rows(**{annotation_name: func(self, *args, **kwargs)})
+            wrapper.mt_prop_meta['annotated'] += 1
+            return self
+
+        wrapper.mt_prop_meta = {
+            'annotated': 0,  # Counter for number of times called. Should only be 0 or 1.
```
[the remark this answers is not in the document]
Comment (knguyen142) [reply luigi_pipeline/lib/model/base_mt_schema.py]: This is for bookkeeping (helpful in our tests to guarantee that), but in the future something could be added to log a warning if it goes over 1 (didn't add it yet because I didn't think it was needed).

File: luigi_pipeline/lib/model/base_mt_schema.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
+from abc import abstractmethod
+from inspect import getmembers, isfunction
+from functools import wraps
+
+
+def mt_annotation(annotation=None, fn_require=None):
```
[the remark this answers is not in the document]
Comment (knguyen142) [reply luigi_pipeline/lib/model/base_mt_schema.py]: Interesting, I was deciding between `annotation` and `name` and thought `name` would be confusing, but I see how it could be clearer. I can change.
Yeah I thought of the `mt_row_annotation` as well to be really general, but so far I've only seen row annotations and didn't know if col should even be supported. But if it makes it clearer I can change.

File: luigi_pipeline/lib/model/seqr_mt_schema.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+from hail_scripts.v02.utils.computed_fields import vep
+
+
+class SeqrSchema(BaseMTSchema):
+
+    @mt_annotation(annotation='sortedTranscriptConsequences')
+    def sorted_transcript_consequences(self):
+        return vep.get_expr_for_vep_sorted_transcript_consequences_array(self.mt.vep)
+
+    @mt_annotation(annotation='docId', fn_require='sorted_transcript_consequences')
```
[the remark this answers is not in the document]
Comment (knguyen142) [reply luigi_pipeline/lib/model/seqr_mt_schema.py]: Nice, I knew some of them didn't depend on it but didn't want to look through each one again haha. Sweet, I'll change.

File: luigi_pipeline/lib/model/seqr_mt_schema.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        return variant_id.get_expr_for_xpos(self.mt.locus)
+
+    @mt_annotation(fn_require='sorted_transcript_consequences')
+    def xstop(self):
+        return variant_id.get_expr_for_xpos(self.mt.locus) + hl.len(variant_id.get_expr_for_ref_allele(self.mt))
+
+    @mt_annotation(fn_require='sorted_transcript_consequences')
+    def domains(self):
+        return vep.get_expr_for_vep_protein_domains_set_from_sorted(
+            self.mt.sortedTranscriptConsequences)
```
[the remark this answers is not in the document]
Comment (knguyen142) [reply luigi_pipeline/lib/model/seqr_mt_schema.py]: Hmm, this is where I'd disagree (assuming you mean calling `domains(dependency=`sortedTranscriptConsequences )`), since the dependency should be a property of the annotation itself, not something that should be changed, and should be fixed in the schema. You shouldn't have 2 callers have different dependencies. 

If you mean changing it so it's `@mt_annotation(annotation_require='sortedTranscriptConsequences')`, I agree that would be better, but would be more complex since it require the explicit mapping between annotation and function. This wouldn't be impossible to do, but I think it would add unnecessary complexity for now since in practice, I think the function name should be the annotation name (ours is weird since I have to support legacy stuff).

File: luigi_pipeline/lib/model/base_mt_schema.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nawatts) [reply luigi_pipeline/lib/model/base_mt_schema.py]: I guess row annotations as the default makes sense even if we later add `mt_column_annotation`.

File: luigi_pipeline/lib/model/base_mt_schema.py (hunk 1 above)
Comment (knguyen142) [reply luigi_pipeline/lib/model/base_mt_schema.py]: Yeah, I thought to add a check to make sure it's a method of the class, but figured both would fail with essentially the same error message. I'll add an informative exception to be explicit.

File: luigi_pipeline/lib/model/seqr_mt_schema.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nawatts) [reply luigi_pipeline/lib/model/seqr_mt_schema.py]: I was thinking something like
```python
@mt_annotation(fn_require='sorted_transcript_consequences')
def domains(self, sorted_transcript_consequences):
  return vep.get_expr_for_vep_protein_domains_set_from_sorted(sorted_transcript_consequences)
```

where the `sorted_transcript_consequences` passed to `domains` is `mt.sortedTranscriptConsequences`.

Thus, if the annotation name on `sorted_transcript_consequences` is changed to `foo`, domains doesn't have to be changed to reference `mt.foo`.

File: luigi_pipeline/lib/model/base_mt_schema.py (hunk 2 above)
Comment (knguyen142) [reply luigi_pipeline/lib/model/base_mt_schema.py]: So I thought it might be confusing, but thought that the `Schema` class would make it more obvious that it's adding annotations. Even thought of naming it `AnnotationBuilder` but `Schema` seems more appropriate since it'll do more.

So adding this functionality just provides more flexibility. (e.g. it was really helpful in our test cases, when we just wanted to test derived fields). It also allows the client flexibility if they wanted partials for some reason (so I was thinking if we wanted to, we can explicitly add stages for derived fields, reference data, etc. Let's provide the functionality for now, and remove (or not advertise) if it's too confusing.

File: luigi_pipeline/lib/model/base_mt_schema.py (hunk 3 above)
Comment (knguyen142) [reply luigi_pipeline/lib/model/base_mt_schema.py]: Yeah this is another case for flexibility, when we want partial annotations. So if you only annotate calls to `a` and `b`, it will only return those.

Comment (nawatts) [reply luigi_pipeline/lib/model/base_mt_schema.py]: Ah, I didn't notice the `if fn[1].mt_prop_meta['annotated'] > 0` and thought this would attempt to select all annotations even if they haven't been applied.

File: luigi_pipeline/lib/model/seqr_mt_schema.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (knguyen142) [reply luigi_pipeline/lib/model/seqr_mt_schema.py]: Ahh interesting idea. And so the arg `sorted_transcript_consequences` would be the return value of that function? That would require some explicit memoization mapping, and the implication is the client can override this (not sure if this is a good or bad thing). It's definitely interesting and I can see it's elegance. Let me sleep on it and if we decide to go with that we'll do another PR.

File: luigi_pipeline/lib/model/base_mt_schema.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+        @wraps(func)
+        def wrapper(self, *args, **kwargs):
+            # Called already.
+            if wrapper.mt_prop_meta['annotated'] > 0:
+                return self
+            if fn_require:
+                getattr(self, fn_require)()
+
+            # Annotate and retiurn instance for chaining.
```
[the remark this answers is not in the document]
Comment (knguyen142) [reply luigi_pipeline/lib/model/base_mt_schema.py]: It's funny, I actually saw this after I pushed and was like, maybe no one will notice. Nice catch!

File: luigi_pipeline/lib/model/seqr_mt_schema.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nawatts) [reply luigi_pipeline/lib/model/seqr_mt_schema.py]: I don't think it would be difficult to implement. Store the return value of `getattr(self, fn_require.__name__)()` and pass it to `func` in `wrapper`.

> That would require some explicit memoization mapping, and the implication is the client can override this (not sure if this is a good or bad thing)
Yeah, I think the difficult part is how to handle arguments when calling individual annotation methods.

[the remark this answers is not in the document]
Comment (knguyen142) [reply luigi_pipeline/lib/model/seqr_mt_schema.py]: Right, because the function expects the arg to be an mt root, but the caller would not have access to that, so we'd have to make it optional e.g. 

`def domains(self, sorted_transcript_consequences=None):`

However, one we actually need to pass the value, we'd have to know the name of the named argument? This also makes it weird because we want to support the ability for the caller to pass in its own vars to the function (e.g. config vars, or other mt's to join).

**Comment by knguyen142** [comment]
@tpoterba Got it, thanks Tim. As long as it's not like tens of thousands of objects copied over each time, I think we should be good performance (in memory) and memory (predictable allocation) wise. The optimization would definitely help, and I'm sure there are some immutable packages out there that make the copy on write logic easier.

**Comment by hanars** [comment]
I like this approach! I think it would be nice if was extended for the reference data fields (like pop frequencies and predictor scores) as well because if I were just coming into this codebase trying to figure where a field came from I would assume it would be in the schema and then when be sad when I couldn't find it there

**Comment by knguyen142** [comment]
@hanars yeah the reference data can use a lot of TLC haha. We'll evaluate if it's run and changed enough to warrant a rewrite though.

**Comment by hanars** [comment]
> @hanars yeah the reference data can use a lot of TLC haha. We'll evaluate if it's run and changed enough to warrant a rewrite though.

I didn't necessarily mean as part of this branch, I more meant in response to "Do you find yourself having a hard time tracking all of the seqr annotations across a long pipeline and from many different sources?" and the answer is a resounding yes but I think this only half-solves the problem since this schema only includes half the annotations that make it all the way downstream.

**Comment by knguyen142** [comment]
Yeah we'll have a combined reference data schema that validates the
annotations when we load it in here, but I thought you were referring
having individual ref data schemas for when we generate it.
